### ISSUE TITLE
Use Word8 for Rank/Suit

### ISSUE THREAD

**Issue by santiweight** [santiweight/poker-base#10]
Pattern Synonyms do the job. Not sure whether:
 1. This actually improves runtime memory/speed characteristics
 2. Whether it's worth making the library slightly more confusing if the benefit is minor

Either way - not a priority - we should leave this for when someone can answer all these questions first.

**Comment by santiweight** [comment]
The same applies potentially to Card and Hand.

**Comment by tonyday567** [comment]
https://github.com/tonyday567/poker-fold/issues/2 is related

Going through the refactor on poker-fold, I reflected on this issue and formed opinions on this.

There's something nice and very grounding about an Ace being an Ace, and not a `WrappedRank 12`. However much you try and hide that detail, library users are going to have to deal with it.

Having a wrapped Word to convert to is about the same as having an enum instance, but with more responsibility. I bugged out on a few of my enum instances along the way - it's very easy to forget to check whether they work like you expect.

For my use case, arrayification and storage is important, and it was easy to just have an isomorph handy for each poker type so this could happen. Whether you HAVE to convert to a Word to store/unpack/bulk calculate using the base types is still an open question, but it was easy to move Word8's around.

**Comment by santiweight** [comment]
Yeah I agree - let's not toy around with "easy-to-use" implementations that are impossible to use in production - it defeats the point. Instead let's just do our best to keep the implementation detail hidden from sight:

```
newtype Suit = Suit Word8

{-# COMPLETE Club, Heart, Diamond, Suit
pattern Club, Heart, Diamond, Spade :: Suit
pattern Club = Suit 1
pattern Club = 2
...
```

I wonder whether we can ACTUALLY hide the implementation details permanently too?

```
instance Show Suit where
  show Club = "Club"
  ...
```

This is really something I think Haskell can give us without false promises - truly cost-free abstractions. I think the library might be barely harder to use, with the only issue I foresee being that the haddock might be a little confusing. Also for very-speed-sensitive use-cases you might see implementation details - but that seems like a plus to me...

So just to be clear - I think we are on the same page here: Word8 is the future of Rank/Suit and we can do the pattern synonyms trick to keep the details hidden. @tonyday567 

I can take this one on  unless you want to take it :)

**Comment by santiweight** [comment]
Well in that case, next time I get the chance, I'll do a translation to Word8 and see how it comes out. I'm actually quite hopeful personally... Then you can take a look and decide whether you hate it :)
